### PR TITLE
Fix/staking stake

### PR DESCRIPTION
**Pull request by jameshowlett977**
https://sovryn.atlassian.net/browse/SOV-1230

### CODE REVIEW

**Review by ororopickpocket — APPROVED** [review]

File: tests/staking/StakingTest.js [hunk 1]
```diff
@@ -236,6 +251,435 @@ contract("Staking", (accounts) => {
...
+
+        // if stakeFor is a vesting contract, getPriorVestingStakeByDate returns the increased vesting stake for stakeFor and until
+        it("should update stake for vesting contracts", async () => {
+            let user = accounts[0];
+            let lockedDate = kickoffTS.add(new BN(TWO_WEEKS).mul(new BN(2)));
+            let amount = new BN(1000);
+            await token.transfer(user, amount);
+            await token.approve(staking.address, amount, { from: user });
+
+            await staking.addContractCodeHash(user);
```
Comment (tjcloa) [inline tests/staking/StakingTest.js:663]: in a normal flow vesting contract stakes for a user's address which differs from the vesting contract  
the test covers a sort of a corner case when a vesting contract stakes for itself which probably should revert

File: contracts/governance/Staking/modules/shared/StakingShared.sol [hunk 2]
```diff
@@ -70,6 +70,7 @@ contract StakingShared is StakingStorageShared, SafeMath96 {
         // abstract
     }
 
+    //SHOULDN'T BE USED, CAN BE EASILY BROKEN USING CONTRACT
```
Comment (tjcloa) [inline contracts/governance/Staking/modules/shared/StakingShared.sol:73]: why not just remove the function to avoid misuse?

**Review by tjcloa — COMMENTED** [review]
pls see inline comments  
also as a general suggestion 
in all tests added, pls consider fuzzy testing approach - validating staking with all possible combinations of msg.sender, stakeFor and delegatee - equal/not equal and zero address combinations (except for ad-hoc tests validating specific values like only stakeFor equals to zero etc.)

File: contracts/governance/Staking/modules/shared/StakingShared.sol (hunk 2 above)
Comment (kostind) [reply contracts/governance/Staking/modules/shared/StakingShared.sol:73]: The answer is quite simple. 
Don't you consider this function can be used in another places?
It's out of scope for this task, should be done as a separate task.

File: tests/staking/StakingTest.js (hunk 1 above)
Comment (kostind) [reply tests/staking/StakingTest.js:663]: We already have tests when msg.sender and stakeFor are different.

File: tests/staking/StakingTest.js [hunk 3]
```diff
@@ -236,6 +251,435 @@ contract("Staking", (accounts) => {
...
+        // if stakeFor is the 0 address, the tokens are staked for the msg.sender
+        // if delegatee is the 0 address, the voting power is delegated to stakeFor
+        it("should change stakeFor and delegatee if zero addresses", async () => {
+            let user = accounts[0];
+            let lockedDate = kickoffTS.add(new BN(TWO_WEEKS).mul(new BN(2)));
+            let amount = new BN(1000);
+            await token.transfer(user, amount);
+            await token.approve(staking.address, amount, { from: user });
+
+            let tx = await staking.stake(amount, lockedDate, ZERO_ADDRESS, ZERO_ADDRESS, {
```
[the remark this answers is not in the document]
Comment (kostind) [reply tests/staking/StakingTest.js:383]: It doesn't make any sense, in this case we should create SEPARATE tests for each case.

File: tests/staking/StakingTest.js [hunk 4]
```diff
@@ -236,6 +251,435 @@ contract("Staking", (accounts) => {
...
+            let user = accounts[0];
+            let lockedDate = kickoffTS.add(new BN(TWO_WEEKS).mul(new BN(2)));
+            let amount = new BN(1000);
+            await token.transfer(user, amount.mul(new BN(2)));
+            await token.approve(stakingWrapperMockup.address, amount.mul(new BN(2)), {
+                from: user,
+            });
+
+            await expectRevert(
+                stakingWrapperMockup.stake2times(amount, lockedDate, user, user, { from: user }),
```
[the remark this answers is not in the document]
Comment (kostind) [reply tests/staking/StakingTest.js:297]: `function _notSameBlockAsStakingCheckpoint(uint256 lockDate, address stakeFor) internal view {
        uint32 nCheckpoints = numUserStakingCheckpoints[stakeFor][lockDate];
        bool notSameBlock =
            userStakingCheckpoints[stakeFor][lockDate][nCheckpoints - 1].fromBlock != block.number;
        require(notSameBlock, "cannot be mined in the same block as last stake"); // S20
    }`

It doesn't matter if we have the same or different transactions date in this case.

[the remark this answers is not in the document]
Comment (kostind) [reply tests/staking/StakingTest.js:297]: Test checks exactly what it NEED to test.

File: tests/staking/StakingTest.js [hunk 5]
```diff
@@ -236,6 +246,347 @@ contract("Staking", (accounts) => {
...
+        //     await network.provider.send("evm_setAutomine", [false]);
+        //     await network.provider.send("evm_setIntervalMining", [10000]);
+        //     await staking.stake(amount, lockedDate, user, user, {from: user});
+        //
+        //     await expectRevert(staking.stake(amount, lockedDate, user, user, {from: user}), "cannot be mined in the same block as last stake");
+        // });
+
+        it("should fail if paused", async () => {
+            await staking.freezeUnfreeze(true);
+            await expectRevert(staking.stake(0, 0, ZERO_ADDRESS, ZERO_ADDRESS), "paused");
```
[the remark this answers is not in the document]
Comment (kostind) [reply tests/staking/StakingTest.js:305]: It doesn't make sense:
1) We check exact error message.
2) The check in modifier will be executed before validating input.
`modifier whenNotPaused() {
        require(!paused, "paused"); // SS03
        _;
    }`

**Comment by kostind** [comment]
I marked all tests I'm going to fix.
Thank you a lot for finding misleading texts :-)

File: tests/staking/StakingTest.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:305]: It doesn't make sense:
1. We check exact error message.
2. The check in modifier will be executed before validating input.
modifier whenNotPaused() { require(!paused, "paused"); // SS03

File: tests/staking/StakingTest.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:297]: `function` _notSameBlockAsStakingCheckpoint(uint256 lockDate, address stakeFor) internal view { uint32 nCheckpoints = numUserStakingCheckpoints[stakeFor][lockDate]; bool notSameBlock = userStakingCheckpoints[stakeFor][lockDate][nCheckpoints - 1].fromBlock != block.number; require(notSameBlock, "cannot be mined in the same block as last stake"); // S20 }`

It doesn't matter if we have the same or different transaction date in this case.
Test checks exactly what it NEEDS to test.

File: tests/staking/StakingTest.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:383]: It doesn't make any sense, in this case we should create a SEPARATE tests for each case.

File: tests/staking/StakingTest.js (hunk 1 above)
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:663]: We already have tests when msg.sender and stakeFor are different.

File: contracts/governance/Staking/modules/shared/StakingShared.sol (hunk 2 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/modules/shared/StakingShared.sol:73]: The answer is quite simple.
Don't you think this function can be used in other places? :-)
It's out of scope for this task, and should be done as a separate task.

**Comment by jameshowlett977** [comment]
I marked all the tests I'm going to fix.
Thank you a lot for finding misleading texts :-)

File: contracts/governance/Staking/modules/shared/StakingShared.sol (hunk 2 above)
Comment (tjcloa) [reply contracts/governance/Staking/modules/shared/StakingShared.sol:73]: ok. then pls restate the comment to add more clarity. otherwise it shouldn't be used but in fact it is.

File: tests/staking/StakingTest.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tjcloa) [reply tests/staking/StakingTest.js:383]: the comments above the test say that it is required to test stakeFor == 0 and delegatee == 0 separately, no?

[the remark this answers is not in the document]
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:383]: No, it a list of what we need to test in current test. Separately or not it doesn't matter in current context.

File: contracts/governance/Staking/modules/shared/StakingShared.sol (hunk 2 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/modules/shared/StakingShared.sol:73]: We will have a separate story for it.

File: tests/staking/StakingTest.js [hunk 6]
```diff
@@ -236,6 +251,435 @@ contract("Staking", (accounts) => {
...
+            let user = accounts[0];
+            let delegatee1 = accounts[1];
+            let delegatee2 = accounts[2];
+            let lockedDate = kickoffTS.add(new BN(TWO_WEEKS).mul(new BN(2)));
+            let amount = new BN(1000);
+            await token.transfer(user, amount);
+            await token.approve(staking.address, amount, { from: user });
+
+            await expectRevert(
+                staking.stake(amount, lockedDate, delegatee1, delegatee2, { from: user }),
```
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:281]: We already have such test: "should stake tokens for user with delegation".

File: tests/staking/StakingTest.js [hunk 7]
```diff
@@ -236,6 +251,435 @@ contract("Staking", (accounts) => {
...
+            let tx = await staking.stake(amount, invalidLockDate, ZERO_ADDRESS, ZERO_ADDRESS, {
+                from: user,
+            });
+            let userBalanceAfter = await token.balanceOf(user);
+            let stakingBalanceAfter = await token.balanceOf(staking.address);
+
+            expect(userBalanceBefore.sub(userBalanceAfter)).to.be.equal(amount);
+            expect(stakingBalanceAfter.sub(stakingBalanceBefore)).to.be.equal(amount);
+
+            await expectEvent.inTransaction(
```
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply tests/staking/StakingTest.js:335]: I don't think we need to check in each tests case that were already checked, 
but ok :-)
